File: gof-design-patterns/cs/P01_Singleton/SingletonExample.c
#include <stdio.h>
#include "Singleton.h"

int main(void) {
    // 通常のnewはできない（C言語では構造体の直接初期化を防ぐ方法がないが、
    // getInstance()を使うことで、同じインスタンスを取得することを保証）
    
    // getInstance()関数でインスタンスを取得
    Singleton* s1 = Singleton_getInstance();
    Singleton* s2 = Singleton_getInstance();
    
    // 同じインスタンスか確認（ポインタの比較）
    printf("s1 == s2: %s\n", (s1 == s2) ? "true" : "false");
    printf("s1 == s2: %d\n", (s1 == s2)); // 1 (true)
    
    // メソッドの呼び出し
    s1->doSomething(s1);
    s2->doSomething(s2);
    
    return 0;
}

File: gof-design-patterns/pythons/P01_Singleton/SingletonExample.py
from Singleton import Singleton

if __name__ == "__main__":
    # Pythonでは、__new__をオーバーライドしているため、
    # 直接インスタンス化しても同じインスタンスが返される
    # ただし、getInstance()を使う方が明示的
    
    # get_instance()メソッドでインスタンスを取得
    s1 = Singleton.get_instance()
    s2 = Singleton.get_instance()
    
    # 直接インスタンス化しても同じインスタンスが返される
    s3 = Singleton()
    
    # 同じインスタンスか確認
    print(f"s1 == s2: {s1 is s2}")  # True
    print(f"s1 == s3: {s1 is s3}")  # True
    print(f"s1 is s2: {s1 is s2}")  # True（is演算子で同一性を確認）
    
    s1.do_something()
